File: trunk/src/google/cacheinvalidation/v2/simple-registration-store.cc
// Copyright 2011 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Simple, map-based implementation of DigestStore.

#include "google/cacheinvalidation/v2/simple-registration-store.h"

#include "google/cacheinvalidation/v2/object-id-digest-utils.h"

namespace invalidation {

void SimpleRegistrationStore::Add(const ObjectIdP& oid) {
  registrations_[ObjectIdDigestUtils::GetDigest(oid, digest_function_)] = oid;
  RecomputeDigest();
}

void SimpleRegistrationStore::Add(const vector<ObjectIdP>& oids) {
  for (size_t i = 0; i < oids.size(); ++i) {
    const ObjectIdP& oid = oids[i];
    registrations_[ObjectIdDigestUtils::GetDigest(oid, digest_function_)] = oid;
  }
  RecomputeDigest();
}

void SimpleRegistrationStore::Remove(const ObjectIdP& oid) {
  registrations_.erase(ObjectIdDigestUtils::GetDigest(oid, digest_function_));
  RecomputeDigest();
}

void SimpleRegistrationStore::Remove(const vector<ObjectIdP>& oids) {
  for (size_t i = 0; i < oids.size(); ++i) {
    const ObjectIdP& oid = oids[i];
    registrations_.erase(ObjectIdDigestUtils::GetDigest(oid, digest_function_));
  }
  RecomputeDigest();
}

void SimpleRegistrationStore::RemoveAll(vector<ObjectIdP>* oids) {
  for (map<string, ObjectIdP>::const_iterator iter = registrations_.begin();
       iter != registrations_.end(); ++iter) {
    oids->push_back(iter->second);
  }
  registrations_.clear();
  RecomputeDigest();
}

bool SimpleRegistrationStore::Contains(const ObjectIdP& oid) {
  return registrations_.find(
      ObjectIdDigestUtils::GetDigest(oid, digest_function_)) !=
      registrations_.end();
}

void SimpleRegistrationStore::GetElements(
    const string& oid_digest_prefix, int prefix_len,
    vector<ObjectIdP>* result) {
  // We always return all the registrations and let the Ticl sort it out.
  for (map<string, ObjectIdP>::iterator iter = registrations_.begin();
       iter != registrations_.end(); ++iter) {
    result->push_back(iter->second);
  }
}

void SimpleRegistrationStore::RecomputeDigest() {
  digest_ = ObjectIdDigestUtils::GetDigest(
      registrations_, digest_function_);
}

}  // namespace invalidation

File: trunk/src/java/com/google/ipc/invalidation/ticl/SimpleRegistrationStore.java
/*
 * Copyright 2011 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.ipc.invalidation.ticl;

import com.google.ipc.invalidation.common.DigestFunction;
import com.google.ipc.invalidation.common.ObjectIdDigestUtils;
import com.google.ipc.invalidation.util.Bytes;
import com.google.ipc.invalidation.util.InternalBase;
import com.google.protos.ipc.invalidation.ClientProtocol.ObjectIdP;

import java.util.ArrayList;
import java.util.Collection;
import java.util.SortedMap;
import java.util.TreeMap;

/**
 * Simple, map-based implementation of {@link DigestStore}.
 *
 */
class SimpleRegistrationStore extends InternalBase implements DigestStore<ObjectIdP> {

  /** All the registrations in the store mapped from the digest to the Object Id. */
  private final SortedMap<Bytes, ObjectIdP> registrations = new TreeMap<Bytes, ObjectIdP>();

  /** The function used to compute digests of objects. */
  private final DigestFunction digestFunction;

  /** The memoized digest of all objects in registrations. */
  private Bytes digest;

  SimpleRegistrationStore(DigestFunction digestFunction) {
    this.digestFunction = digestFunction;
    recomputeDigest();
  }

  @Override
  public void add(ObjectIdP oid) {
    registrations.put(ObjectIdDigestUtils.getDigest(oid, digestFunction), oid);
    recomputeDigest();
  }

  @Override
  public void add(Collection<ObjectIdP> oids) {
    for (ObjectIdP oid : oids) {
      registrations.put(ObjectIdDigestUtils.getDigest(oid, digestFunction), oid);
    }
    recomputeDigest();
  }

  @Override
  public void remove(ObjectIdP oid) {
    registrations.remove(ObjectIdDigestUtils.getDigest(oid, digestFunction));
    recomputeDigest();
  }

  @Override
  public void remove(Collection<ObjectIdP> oids) {
    for (ObjectIdP oid : oids) {
      registrations.remove(ObjectIdDigestUtils.getDigest(oid, digestFunction));
    }
    recomputeDigest();
  }

  @Override
  public Collection<ObjectIdP> removeAll() {
    Collection<ObjectIdP> result = new ArrayList<ObjectIdP>(registrations.values());
    registrations.clear();
    recomputeDigest();
    return result;
  }

  @Override
  public boolean contains(ObjectIdP oid) {
    return registrations.containsKey(ObjectIdDigestUtils.getDigest(oid, digestFunction));
  }

  @Override
  public int size() {
    return registrations.size();
  }

  @Override
  public byte[] getDigest() {
    return digest.getByteArray();
  }

  @Override
  public Collection<ObjectIdP> getElements(byte[] oidDigestPrefix, int prefixLen) {
    // We always return all the registrations and let the Ticl sort it out.
    return registrations.values();
  }

  /** Recomputes the digests over all objects and sets {@code this.digest}. */
  private void recomputeDigest() {
    this.digest = ObjectIdDigestUtils.getDigest(registrations.keySet(), digestFunction);
  }
}
